finality-aleph/src/data_io/proposal.rs

```diff
@@ -0,0 +1,174 @@
+#[derive(Clone, Debug, Encode, Decode)]
+pub struct AlephUnvalidatedProposal<B: BlockT> {
+    pub branch: Vec<B::Hash>,
+    pub number: NumberFor<B>,
+}
+
+// Need to be implemented manually, as deriving does not work (`BlockT` is not `Hash`).
+impl<B: BlockT> Hash for AlephUnvalidatedProposal<B> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.branch.hash(state);
+        self.number.hash(state);
+    }
+}
+
+// Clippy does not allow deriving PartialEq when implementing Hash manually
+impl<B: BlockT> PartialEq for AlephUnvalidatedProposal<B> {
+    fn eq(&self, other: &Self) -> bool {
+        self.number.eq(&other.number) && self.branch.eq(&other.branch)
+    }
+}
+
+impl<B: BlockT> Eq for AlephUnvalidatedProposal<B> {}
+
+impl<B: BlockT> AlephUnvalidatedProposal<B> {
+    pub(crate) fn new(branch: Vec<B::Hash>, block_number: NumberFor<B>) -> Self {
+        AlephUnvalidatedProposal {
+            branch,
+            number: block_number,
+        }
+    }
+
+    pub(crate) fn validate_bounds(
+        &self,
+        session_boundaries: &SessionBoundaries<B>,
+    ) -> Option<AlephProposal<B>> {
+        if self.branch.len() > MAX_DATA_BRANCH_LEN as usize {
+            return None;
+        }
+        if self.branch.is_empty() {
+            return None;
+        }
+        if self.number < <NumberFor<B>>::saturated_from(self.branch.len()) {
+            return None;
+        }
+
+        let bottom_block = self.number - <NumberFor<B>>::saturated_from(self.branch.len() - 1);
+        let top_block = self.number;
+        if session_boundaries.first_block() <= bottom_block
+            && top_block <= session_boundaries.last_block()
+        {
+            return Some(AlephProposal {
+                branch: self.branch.clone(),
+                number: self.number,
+            });
+        }
+        None
+    }
+}
+
+/// A version of AlephUnvalidatedProposal that has been initially validated and fits
+/// within session bounds.
+#[derive(Clone, Debug, Encode, Decode)]
+pub struct AlephProposal<B: BlockT> {
+    pub branch: Vec<B::Hash>,
+    pub number: NumberFor<B>,
+}
+
+// Need to be implemented manually, as deriving does not work (`BlockT` is not `Hash`).
+impl<B: BlockT> Hash for AlephProposal<B> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.branch.hash(state);
+        self.number.hash(state);
+    }
+}
+
+// Clippy does not allow deriving PartialEq when implementing Hash manually
+impl<B: BlockT> PartialEq for AlephProposal<B> {
+    fn eq(&self, other: &Self) -> bool {
+        self.number.eq(&other.number) && self.branch.eq(&other.branch)
+    }
+}
+
+impl<B: BlockT> Eq for AlephProposal<B> {}
+
+impl<B: BlockT> AlephProposal<B> {
+    /// Outputs the highest block in the branch.
+    pub(crate) fn top_block(&self) -> BlockHashNum<B> {
+        (
+            *self
+                .branch
+                .last()
+                .expect("cannot be empty for correct data"),
```

Comment:
Ehh, then I need to make sure this is not used in serialization + 100 other annoying consequences that I cannot think of now. Introducing `UnvalidatedAlephProposal` was already the peak of type safety for me :)

Reply:
This doesn't ever get serialized, does it? Hm.

Anyway, I guess this can be left for the future as well, the `expect`s are annoying, but they always will be so eventually someone will fix this.